internal/fields/validate.go

```diff
@@ -0,0 +1,196 @@
+		}
+		fields = append(fields, u...)
+	}
+	return &Validator{
+		schema: fields,
+	}, nil
+}
+
+// ValidateDocumentBody validates the provided document body.
+func (v *Validator) ValidateDocumentBody(body json.RawMessage) error {
+	var c common.MapStr
+	err := json.Unmarshal(body, &c)
+	if err != nil {
+		return errors.Wrap(err, "unmarshalling document body failed")
+	}
+	return v.validateMapElement("", c)
+}
+
+// ValidateDocumentMap validates the provided document as common.MapStr.
+func (v *Validator) ValidateDocumentMap(body common.MapStr) error {
+	return v.validateMapElement("", body)
+}
+
+func (v *Validator) validateMapElement(root string, elem common.MapStr) error {
+	var err error
+	for name, val := range elem {
+		key := strings.TrimLeft(root+"."+name, ".")
+
+		switch val.(type) {
+		case []map[string]interface{}:
+			for _, m := range val.([]map[string]interface{}) {
+				err = v.validateMapElement(key, m)
+				if err != nil {
+					return err
+				}
+			}
+		case map[string]interface{}:
+			err = v.validateMapElement(key, val.(map[string]interface{}))
+			if err != nil {
+				return err
+			}
+		default:
+			err = v.validateElementFormat(key, val)
+			if err != nil {
+				return err
+			}
+			break
+		}
+	}
+	return nil
+}
+
+func (v *Validator) validateElementFormat(key string, val interface{}) error {
+	if key == "" {
+		return nil // root key is always valid
+	}
+
+	if skipValidationForField(key) {
+		return nil // generic field, let's skip validation for now
+	}
+
+	definition := findElementDefinitionInSlice("", key, v.schema)
+	if definition == nil {
+		return fmt.Errorf(`field "%s" is not defined`, key)
+	}
+
+	err := parseElementValue(key, *definition, val)
+	if err != nil {
+		return errors.Wrap(err, "parsing field value failed")
+	}
+	return nil
+}
+
+func skipValidationForField(key string) bool {
+	return isFieldFamilyMatching("agent", key) ||
+		isFieldFamilyMatching("elastic_agent", key)
+}
+
+func isFieldFamilyMatching(family, key string) bool {
+	return key == family || strings.HasPrefix(key, family+".")
+}
+
+func findElementDefinitionInSlice(root, searchedKey string, fieldDefinitions []fieldDefinition) *fieldDefinition {
+	for _, def := range fieldDefinitions {
+		key := strings.TrimLeft(root+"."+def.Name, ".")
+		if compareKeys(key, def, searchedKey) {
+			return &def
+		}
+
+		if len(def.Fields) == 0 {
+			continue
+		}
+
+		fd := findElementDefinitionInSlice(key, searchedKey, def.Fields)
+		if fd != nil {
+			return fd
+		}
+	}
+	return nil
+}
+
+func compareKeys(key string, def fieldDefinition, searchedKey string) bool {
+	k := strings.ReplaceAll(key, ".", "\\.")
+	k = strings.ReplaceAll(k, "*", "[^.]+")
+
+	// Workaround for potential geo_point, as "lot" and "lat" fields are not present in field definitions.
+	if def.Type == "geo_point" {
+		k += "\\.(lon|lat)"
+	}
+
+	k = fmt.Sprintf("^%s$", k)
+	matched, err := regexp.MatchString(k, searchedKey)
+	if err != nil {
+		panic(err)
```

Comment:
Since we're `panic`ing here, maybe add a bit more context with `errors.Wrap`?

Reply:
Fixed.